src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -6,54 +6,208 @@
+  int16_t state = RADIOLIB_ERR_UNKNOWN;
+  Module* mod = this->phyLayer->getMod();
+  
+  // if after (at) ADR_ACK_LIMIT frames no RekeyConf was received, revert to Join state
+  if(this->fCntUp == (1UL << this->adrLimitExp)) {
+    state = this->getMacPayload(RADIOLIB_LORAWAN_MAC_REKEY, this->fOptsUp, this->fOptsUpLen, NULL, RADIOLIB_LORAWAN_UPLINK);
+    if(state == RADIOLIB_ERR_NONE) {
+      this->clearSession();
+    }
+  }
+
+  // if not joined, don't do anything
+  if(!this->isActivated()) {
+    return(RADIOLIB_ERR_NETWORK_NOT_JOINED);
+  }
+  
+  if(this->adrEnabled) {
+    this->adrBackoff();
+  }
+
+  // check if the requested payload + fPort are allowed, also given dutycycle
+  uint8_t totalLen = lenUp + this->fOptsUpLen;
+  state = this->isValidUplink(&totalLen, fPort);
+  RADIOLIB_ASSERT(state);
+
+  // in case of TS009, a payload that is too long may have gotten clipped, 
+  // so recalculate the actual payload length
+  // (outside of TS009, a payload that is too long throws an error)
+  lenUp = totalLen - this->fOptsUpLen;
+
+  // the first 16 bytes are reserved for MIC calculation blocks
+  size_t uplinkMsgLen = RADIOLIB_LORAWAN_FRAME_LEN(lenUp, this->fOptsUpLen);
+  #if RADIOLIB_STATIC_ONLY
+  uint8_t uplinkMsg[RADIOLIB_STATIC_ARRAY_SIZE];
+  #else
+  uint8_t* uplinkMsg = new uint8_t[uplinkMsgLen];
+  #endif
+  
+  // build the encrypted uplink message
+  this->composeUplink(dataUp, lenUp, uplinkMsg, fPort, isConfirmed);
+
+  // reset Time-on-Air as we are starting new uplink sequence
+  this->lastToA = 0;
+
+  // repeat uplink+downlink up to 'nbTrans' times (ADR)
+  uint8_t trans = 0;
+  for(; trans < this->nbTrans; trans++) {
+
+    // keep track of number of hopped channels
+    uint8_t numHops = this->maxChanges;
+
+    // number of additional CAD tries
+    uint8_t numBackoff = 0;
+    if(this->backoffMax) {
+      numBackoff = this->phyLayer->random(1, this->backoffMax + 1);
+    }
+
+    do {
+      // select a pair of Tx/Rx channels for uplink+downlink
+      this->selectChannels();
+
+      // generate and set uplink MIC (depends on selected channel)
+      this->micUplink(uplinkMsg, uplinkMsgLen);
+
+    // if CSMA is enabled, repeat channel selection & encryption up to numHops times
+    } while(this->csmaEnabled && numHops-- > 0 && !this->csmaChannelClear(this->difsSlots, numBackoff));
+    
+    // send it (without the MIC calculation blocks)
+    state = this->transmitUplink(&this->channels[RADIOLIB_LORAWAN_UPLINK],
+                                &uplinkMsg[RADIOLIB_LORAWAN_FHDR_LEN_START_OFFS], 
+                                (uint8_t)(uplinkMsgLen - RADIOLIB_LORAWAN_FHDR_LEN_START_OFFS));
+    if(state != RADIOLIB_ERR_NONE) {
+      #if !RADIOLIB_STATIC_ONLY
+      delete[] uplinkMsg;
+      #endif
+      RADIOLIB_ASSERT(state);
```

Comment:
This macro here effectively adds another if condition, which is not necessary - we already know it will fail, so just `return(state)` is better here.